internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +87,114 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
+			}
+			change.Abort()
+			m.deleteHealthInfo(details.Name)
+			shouldEnsure = true
+		}
+	}
 
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
+	// Also find checks that are new (in new plan but not in old one).
+	for _, config := range newPlan.Checks {
+		if oldPlan.Checks[config.Name] == nil {
+			newOrModified[config.Name] = true
+		}
+	}
 
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
+	// Start new or modified checks.
+	for _, config := range newPlan.Checks {
+		if newOrModified[config.Name] {
+			performCheckChange(m.state, config)
+			m.updateHealthInfo(config, 0)
+			shouldEnsure = true
+		}
+	}
+	if !m.ensureDone.Load() {
+		// Can't call EnsureBefore before Overlord.Loop is running (which will
+		// call m.Ensure for the first time).
+		return
+	}
+	if shouldEnsure {
+		m.state.EnsureBefore(0) // start new tasks right away
 	}
-	// Wait for all context cancellations to propagate and allow
-	// each goroutine to cleanly exit.
-	m.wg.Wait()
-
-	// Set the size of the next wait group
-	m.wg.Add(len(p.Checks))
-
-	// Then configure and start new checks.
-	checks := make(map[string]*checkData, len(p.Checks))
-	for name, config := range p.Checks {
-		ctx, cancel := context.WithCancel(context.Background())
-		check := &checkData{
-			config:  config,
-			checker: newChecker(config, p),
-			ctx:     ctx,
-			cancel:  cancel,
-			action:  m.callFailureHandlers,
+}
+
+func (m *CheckManager) changeStatusChanged(chg *state.Change, old, new state.Status) {
+	// Always acquire locks in same order (state lock, then plan lock).
+	// The state engine has already acquired the state lock at this point.
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+
+	shouldEnsure := false
+	switch {
+	case chg.Kind() == performCheckKind && new == state.ErrorStatus:
+		details := mustGetCheckDetails(chg)
+		config, inPlan := plan.Checks[details.Name]
+		if details.Proceed && inPlan {
+			recoverCheckChange(m.state, config, details.Failures)
+			shouldEnsure = true
+		}
+
+	case chg.Kind() == recoverCheckKind && new == state.DoneStatus:
+		details := mustGetCheckDetails(chg)
+		config, inPlan := plan.Checks[details.Name]
+		if details.Proceed && inPlan {
+			performCheckChange(m.state, config)
+			shouldEnsure = true
 		}
-		checks[name] = check
-		go func() {
-			defer m.wg.Done()
-			check.loop()
-		}()
 	}
-	m.checks = checks
+
+	if shouldEnsure {
+		m.state.EnsureBefore(0) // start new tasks right away
+	}
+}
+
+func mustGetCheckDetails(change *state.Change) checkDetails {
+	tasks := change.Tasks()
+	if len(tasks) != 1 {
+		panic(fmt.Sprintf("internal error: %s change %s should have one task", change.Kind(), change.ID()))
+	}
+	var details checkDetails
+	err := tasks[0].Get(checkDetailsAttr, &details)
+	if err != nil {
+		panic(fmt.Sprintf("internal error: cannot get %s change %s check details: %v", change.Kind(), change.ID(), err))
+	}
+	return details
```

Comment:
Could it remove a couple of lines of code if you place the ```checkDetailsAttr``` at the change level (```change.Get/Set```) ?

Reply:
I realise it's much of a muchness in this case, as we always have one task per change, but given that each handler takes the Task and not the Change object, it seems "righter" to do it this way. (I'd have to do the `task.Change()` indirection in the handler, which doesn't feel great.) So I think I'll leave this one as is.